### DISCUSSION TITLE
On the Linux system, how to view the content output by Debug.WriteLine?

### DISCUSSION THREAD

**Discussion by pedoc** [dotnet/runtime#86743] [Q&A]
On windows we have many tools to view what Debug.WriteLine outputs, but how to view it on linux system? Or will this become a nop operation on linux?

thanks

**Comment by hez2010** [comment]
You need to attach a debugger to the process to see the output.

**Reply by pedoc** [reply]
> You need to attach a debugger to the process to see the output.

Yes, I know it's one of the options.
Similarly, there are some independent tools that can view Debug log output without a debugger, such as https://github.com/CobaltFusion/DebugViewPP

I found out that Debug.WriteLine ends up calling OutputDebugString in win32API, maybe linux doesn't have something similar, so it's a nop under linux?

There is a syslog api available under Linux, they are not exactly equal, but there is some help.

hope this is useful for others.

**Reply by kevingosse** [reply]
Interestingly, it looks like there are internal hooks available for `Debug.Write` APIs: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Diagnostics/DebugProvider.cs#L111-L113

It should be possible to write a lib injected using a startup hook (https://github.com/dotnet/runtime/blob/main/docs/design/features/host-startup-hook.md) that injects it's own Debug handler using reflection. Then that lib can write to syslog or whatever interesting sink we have. Maybe I'll give it a try, it could be a nice side-project.

**Reply by jasper-d** [reply]
> that injects it's own Debug handler using reflection.

Isn't that what Trace.Listeners is for https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.trace.listeners?view=net-7.0 or do you have something different in mind?

**Reply by kevingosse** [reply]
I could be wrong, but I think trace listeners only allow to listen to `Trace.Write`? Or can they listen to `Debug.Write` as well?

**Reply by kevingosse** [reply]
I should have read the documentation before replying.

```
The Listeners collection is shared by both the Debug and the Trace classes; adding a trace listener to either class adds the listener to both.
```
